proposals/2312-matrix-uri.md

```diff
@@ -356,6 +356,95 @@ Client authors are strongly encouraged to standardise custom query elements
+used primarily on the Web to onboard new users, perhaps with matrix: URIs
+alongside them for the convenience of existing Matrix users.  Within Matrix
+clients, we expect matrix: URIs to be the default.
+
+(One alternative considered was to always route via https://matrix.to if there
+is no registered matrix URL handler - however, on reflection, the additional
+UX friction of indirecting everything via matrix.to outweighs the benefits of
+the simpler URI-handling implementation, and makes matrix.to an unnecessary
+single point of failure for onboarding.)
+
+This section and the one below effectively supersedes
+[MSC2644](https://github.com/matrix-org/matrix-doc/pull/2644).
+
+##### Indicating that a link requires a given client or homeserver
+
+Ideally all links would work with all clients and all homeservers. However,
+it's possible that the link author might want to indicate that the link should
+be opened in a specific client or homeserver - e.g. if it requires
+experimental functionality only implemented on that client or server, or if
+the server has been somehow locked to require a specific client.
+
+* `force_client` specifies that a given client must be used to successfully
+follow this link.  If the user has no registered matrix URI handler, the
+behaviour is the same as `default_client` (but takes precedence over that
+parameter).  If the user has a matrix URI handler registered for a different
+client, then the registered client should instead route the user as if
+onboarding for the first time, treating it like a `default_client` parameter.
+
+* `force_hs` specifies that users must use a given homeserver to successfully
+follow this link.  (This is not to be confused with the `via` parameter, which
+tells your *server* which server to use to join this link).  If the user has
+no registered matrix URI handler, the behaviour is the same as `default_hs`
+(but takes precedence over that parameter).  If the user has a matrix URI
+handler registered, then the registered client should route the user to use the
+given homeserver - either by onboarding them to it via login/register (if the
+client has room to add a new account) or by redirecting them to matrix.to or
+the specified `force_client` or `default_client` destination.
+
+These parameters could also be used as a crass way to link to unfederated
+servers or those on private networks, but this is not recommended, given it
+breaks badly in the face of private federations (i.e. by specifying the use of
+a single required homeserver, users on other servers in the private federation
+will be incorrectly forced to use the specified one).  Handling private
+federations will be left for [MSC3020](https://github.com/matrix-org/matrix-doc/pull/3020)
+
+We deliberately don't use the `authority` section of the URI for this given
+the potential ambiguity, whereas assigning it to a query param makes it much
+clearer and avoids it inappropriately dominating the URI.
```

Comment:
Technically `force_client` quite clearly conveys the meaning of an authority, since the client is your gateway to the closed federation. The point about ambiguity is noted; but given that such a URI leads to a different network, I'm still unsure if the client address "dominating the URI" is a bad or rather good thing. I really appreciated when I could quickly distinguish between the "internal" and "public" URLs just by glancing at its beginning. Determining whether the URI is "internal" by seeking the right query parameter somewhere down the string doesn't seem to be user-friendly - if anything, it's a good medium for phishing.

Reply:
hm, do you mean `force_hs`? if so, it's really not intended that the URI leads to a different network (that's what MSC3020 is for).  It's just saying "your client must connect to this HS for this link to work as intended".  I'm not sure that makes the HS more or less an authority than any other HS in the room though...